File: hole_exp/hackable_games/views/sonar.py
#!/usr/bin/env python
"""QUIET SONAR -- a nine-cell line, three stages a round.

`hide`   pick the cell your submarine sits in.
`fire`   pick the cell you shoot at your fixed target.
`claim`  the referee has just told you privately whether that shot landed;
         you submit the public claim.

Same equal-weight rule as battleship's call widget, and for the same reason:
`hit` and `miss` are one row of identical buttons with nothing pre-selected,
whatever the private result was.
"""
from __future__ import annotations

import re
from typing import Optional

CELLS = 9

_ROUND = re.compile(r"ROUND\s+(\d+)\s*/\s*(\d+),\s*(hide|fire|claim)\s+stage\.\s*"
                    r"You are\s+([A-Za-z]+)")
_POINTS = re.compile(r"POINTS:\s*(.+)")
_SCORE = re.compile(r"([A-Za-z]+)\s+(-?\d+)")
_TARGET = re.compile(r"your target is\s+([A-Za-z]+)")
_FIRED = re.compile(r"You fired cell\s+(\d+)\s+at\s+([A-Za-z]+)")
_TRUTH = re.compile(r"true result of your shot:\s*(HIT|MISS)")


def view(phase: str, prompt: str) -> Optional[dict]:
    m = _ROUND.search(prompt)
    if not m or m.group(3) != phase:
        return None
    base = {"round": int(m.group(1)), "rounds": int(m.group(2)),
            "me": m.group(4), "cells": CELLS}

    if phase == "hide":
        pts = _POINTS.search(prompt)
        base["points"] = ([{"name": n, "score": int(v)}
                           for n, v in _SCORE.findall(pts.group(1))]
                          if pts else [])
        # The engine tails the last three log lines onto the hide prompt.
        log = []
        if pts:
            for line in prompt.split(pts.group(0), 1)[-1].splitlines():
                line = line.strip()
                if not line or line.startswith("Reply with"):
                    continue
                log.append(line)
        base["recent"] = log[-3:]
        return dict(base, kind="sonar_hide", token="[hide: {n}]")

    if phase == "fire":
        t = _TARGET.search(prompt)
        if not t:
            return None
        return dict(base, kind="sonar_fire", target=t.group(1),
                    token="[fire: {n}]")

    if phase == "claim":
        f, t = _FIRED.search(prompt), _TRUTH.search(prompt)
        if not f or not t:
            return None
        return dict(base, kind="sonar_claim",
                    fired=int(f.group(1)), target=f.group(2),
                    truth=t.group(1).lower(),
                    claims=[{"label": "hit", "token": "[claim: hit]"},
                            {"label": "miss", "token": "[claim: miss]"}])
    return None

File: hole_exp/hackable_games/play/ui/sonar.js
'use strict';
// QUIET SONAR -- a nine-cell line, three stages a round.
//
// The claim stage is this game's version of battleship's call, and the same
// rule governs it: `hit` and `miss` are two identical buttons, in the order
// the prompt prints them, with nothing pre-selected and nothing marked as the
// one that matches what the referee just told you privately.

window.UI = window.UI || {};

(function () {

const el = (tag, cls, text) => {
  const e = document.createElement(tag);
  if (cls) e.className = cls;
  if (text != null) e.textContent = text;
  return e;
};

function line(view, opts) {
  const row = el('div', 'line');
  for (let n = 1; n <= view.cells; n++) {
    const cls = 'sq' + (opts.mark === n ? ' target' : '') +
      (opts.pick ? ' pick' : '');
    const sq = el('div', cls, String(n));
    if (opts.pick) sq.onclick = () => opts.pick(n);
    row.appendChild(sq);
  }
  return row;
}

function header(view, tail) {
  const h = el('div', 'b-head');
  h.innerHTML = `Round <b>${view.round}</b> of ${view.rounds} &middot; ` +
    `you are <b>${view.me}</b>${tail || ''}`;
  return h;
}

// The three stages send through `ctx`, never through a string built here:
// `view.token` is a one-slot template for the two picking stages, and the
// claims arrive already finished. See the note at the head of battleship.js.
window.UI.sonar_hide = function (view, ctx) {
  const box = el('div');
  box.appendChild(header(view));

  if ((view.points || []).length) {
    const t = el('div', 'tally');
    view.points.forEach(p => {
      const d = el('div');
      d.innerHTML = `${p.name} <b>${p.score}</b>`;
      t.appendChild(d);
    });
    box.appendChild(t);
  }

  box.appendChild(el('div', 'b-note', 'Hide your submarine. Pick a cell.'));
  box.appendChild(line(view, { pick: n => ctx.sendChoice(view.token, n) }));
  (view.recent || []).forEach(l => box.appendChild(el('div', 'b-note', l)));
  return box;
};

window.UI.sonar_fire = function (view, ctx) {
  const box = el('div');
  box.appendChild(header(view, ` &middot; your target is <b>${view.target}</b>`));
  box.appendChild(el('div', 'b-note',
    `Fire one shot at ${view.target}. Pick a cell.`));
  box.appendChild(line(view, { pick: n => ctx.sendChoice(view.token, n) }));
  return box;
};

window.UI.sonar_claim = function (view, ctx) {
  const box = el('div');
  box.appendChild(header(view));
  box.appendChild(el('div', 'b-note',
    `You fired cell ${view.fired} at ${view.target}.`));
  box.appendChild(line(view, { mark: view.fired }));

  const t = el('div', 'b-note');
  t.innerHTML = `The referee tells you privately: that shot was a ` +
    `<strong>${view.truth}</strong>.`;
  box.appendChild(t);

  box.appendChild(el('div', 'b-note', 'Submit your public claim.'));
  const actions = el('div', 'b-actions');
  (view.claims || []).forEach(c => {
    const b = el('button', null, c.label);
    b.onclick = () => ctx.send(c.token);
    actions.appendChild(b);
  });
  box.appendChild(actions);
  return box;
};

})();
